### DISCUSSION TITLE
Component를 만드는 기준에 대한 논의

### DISCUSSION THREAD

**Discussion by hy57in** [teamsindy20/sobok#35] [React.js]
React를 이용해 프로젝트를 하다보면 한번쯤은 component 분리 기준에 생각해보았을 것입니다.

저는 component의 가장 큰 특징이 "분리"와 "재사용"이라고 생각하는데요,
반복되는 패턴이 두군데 이상 있다면 분리시키는 것이 좋다고 생각합니다.

관련 자료들을 서치해본 결과,
실질적인 코드의 재사용성 또는 코드의 가독성과 명시적인 코드의 모양새를 중점적으로 보는 의견이 가장 많이 엇갈리는 것 같습니다.

혹시 태욱님은 component 분리 기준에 대해 생각해보신 적이 있나요? 🤔

**Comment by hy57in** [comment]
@rmfpdlxmtidl 아 그리구 소복 페이지 설계할 때 HomeLayout과 HomePage을 구분하는 기준은 어떤 것이라 생각하시나요?

**Reply by gwak2837** [reply]
레이아웃은 페이지별 공통된 레이아웃을 분리하는 것이고, 페이지는 페이지별로 `url`이 있는 Next.js 렌더링 기본 단위라고 생각해요.

**Comment by gwak2837** [comment]
결국 개발 속도를 높이는 것이 중요하기 때문에 본인의 추상화 능력(컴포넌트 로직을 계층화할 수 있는 능력)에 따라서 적절히 분리하면 좋을 거 같아요ㅎㅎㅎ

코드 재사용성은 무조건 컴포넌트로 분리하는 것이 좋아요. 하지만 코드 가독성을 위해 일정 단위를 컴포넌트로 분리하면 프로젝트의 구조가 일정 부분 정해지기 때문에, 나중에 유지 보수 및 새로운 기능 추가 시 높은 확률로 기존 구조를 변경해야 하는 불편함이 있을 수 있어요. 본인이 컴포넌트 로직 계층화 및 추상화를 잘 할 수 있으면 유지 보수할 때 추가적인 수정이 별로 없겠지만, 그렇지 않다면 기존 코드 수정이 많이 있을 수 있어요.

이 부분에 대해선 객체 지향의 [SOLID](https://ko.wikipedia.org/wiki/SOLID_(%EA%B0%9D%EC%B2%B4_%EC%A7%80%ED%96%A5_%EC%84%A4%EA%B3%84)) 원칙도 참고하면 좋을 거 같아요~
